### PR TITLE
Resolve NaN, add Ranger, improve (parameter-)efficiency

### PR DESCRIPTION
**Pull request by ClashLuke**

### CODE REVIEW

**Comment by ClashLuke** [comment]
Loss curve with v1.10
image
image
where the long, dense, squiggly line is the model getting into NaN's.\
Those are what this pull request aims to fix. Whether that's with label smoothing, increased weight decay, increased output regularization, or anything else doesn't matter.\
On the topic of output regularization, the "z loss" used in Mesh TensorFlow's cross-entropy implementation originally adds [logsumexp(output)^2*1e-4](https://github.com/tensorflow/mesh/blob/83d1e0f8f7bdd088241e1d7c64a0d914f8ec0ad1/mesh_tensorflow/layers.py#L1100) to the loss. What we do here however is add [output^2*0.01](https://github.com/ClashLuke/JanNet/blob/7cdf8973de0fd2b7ad0cc45d009449836591baf2/src/dataclass.py#L217) to the loss, to increase the stability of the output range. Additionally, we significantly optimized the summation and reduction done in the cross-entropy calculation by only ever adding scalar values rather than huge matrices. While this shouldn't give large speed-ups, I'm expecting some stability gains as there's less intermediate sums.

**Comment by ClashLuke** [comment]
It seems to be a bit more stable now (also reduced the parameters 4x, so now they're at 2x compared to before).
image

**Comment by ClashLuke** [comment]
* it's not rezero
* it's not the optimizer (ranger is worse than Adam here)
* z_loss and label smoothing help (+/-40% runtime till NaN) but don't solve the issue
* it only happens when using both language token and image patches
* it didn't happen without all-attention

At this point, I'm convinced there's an issue in the attention computation. Possible culprits here are:
* too low normalization -> fix division by sqrt
* sofmax across 8000 token doest work -> split sofmax
* learned k and v are initialized poorly -> use randn instead of orthogonal init

**Comment by ClashLuke** [comment]
Using randn intiliazation for the embedding helps slightly, but doesn't solve the NaN issue.
image
Light blue is the previous best (with strong label smoothing and a high z_loss), dark green has the same parameters but uses the custom embedding function of the most recent commit.

**Comment by ClashLuke** [comment]
Regarding NaNs, the actual purpose of this pull request, they seem to still be there at roughly the same step.
image
Red is a run with value clipping, pink added strong label smoothing, and grey added normal embedding initialization.

**Comment by ClashLuke** [comment]
image
Fixing the attention scale factor already helps a bit, so it's most likely related to the attention matrix.

**Comment by ClashLuke** [comment]
image
After fixing the sign of added max value of logsumexp, the runs are stable again.
Dark blue is the previous run with fixed scaling, orange is the initial baseline seen above, and light blue is the new run.
It should be quite obvious that having a z_loss of 0.01 and label smoothing of 0.2 doesn't quite help the model anymore. Still, I'll let the run converge and report back the final accuracies.

Below are plots of the runs without language, which clearly converge significantly better with all-attention and a working softmax implementation.
image
image
